File: pkg/input/mouse.go
package input

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
	"unsafe"

	"golang.org/x/sys/unix"
)

const (
	reportSize = 3
)

// MouseEvent is a decoded mouse packet
type MouseEvent struct {
	Timestamp int64
	Device    string
	Buttons   []string
	DX, DY    int8
}

// direction converts x and y deltas into a human readable description such as
// "up", "down-right", etc. A zero movement returns "none".
func direction(dx, dy int8) string {
	var dir string

	if dy < 0 {
		dir += "up"
	} else if dy > 0 {
		dir += "down"
	}

	if dx < 0 {
		if dir != "" {
			dir += "-"
		}
		dir += "left"
	} else if dx > 0 {
		if dir != "" {
			dir += "-"
		}
		dir += "right"
	}

	if dir == "" {
		dir = "none"
	}

	return dir
}

// String renders a human-friendly representation of the mouse event.
func (e MouseEvent) String() string {
	return fmt.Sprintf("[%d ms] %s: Buttons[%s] Move[%s]",
		e.Timestamp,
		e.Device,
		strings.Join(e.Buttons, ","),
		direction(e.DX, e.DY))
}

type MouseDevice struct {
	Path string
	FD   int
}

func openMouseDevice(path string) (*MouseDevice, error) {
	fd, err := unix.Open(path, unix.O_RDONLY|unix.O_NONBLOCK, 0)
	if err != nil {
		return nil, err
	}
	fmt.Printf("[+] Opened %s (fd=%d)\n", path, fd)
	return &MouseDevice{Path: path, FD: fd}, nil
}

func (m *MouseDevice) Close() {
	if m.FD >= 0 {
		_ = unix.Close(m.FD)
		fmt.Printf("[-] Closed %s\n", m.Path)
		m.FD = -1
	}
}

// StreamMouse returns a channel of MouseEvent.
// It starts goroutines to monitor /dev/input/mouse* and /dev/input/mice.
func StreamMouse() <-chan MouseEvent {
	out := make(chan MouseEvent, 100) // buffered channel

	go func() {
		defer close(out)
		devices := map[string]*MouseDevice{}

		// initial scan
		paths, _ := filepath.Glob("/dev/input/mouse*")
		if _, err := os.Stat("/dev/input/mice"); err == nil {
			paths = append(paths, "/dev/input/mice")
		}
		for _, path := range paths {
			if dev, err := openMouseDevice(path); err == nil {
				devices[path] = dev
			}
		}

		// inotify watch on /dev/input
		inFd, err := unix.InotifyInit()
		if err != nil {
			fmt.Println("inotify init failed:", err)
			return
		}
		defer unix.Close(inFd)

		_, err = unix.InotifyAddWatch(inFd, "/dev/input", unix.IN_CREATE|unix.IN_DELETE)
		if err != nil {
			fmt.Println("inotify addwatch failed:", err)
			return
		}

		for {
			// poll devices + inotify
			var pollfds []unix.PollFd
			for _, dev := range devices {
				if dev.FD >= 0 {
					pollfds = append(pollfds, unix.PollFd{Fd: int32(dev.FD), Events: unix.POLLIN})
				}
			}
			pollfds = append(pollfds, unix.PollFd{Fd: int32(inFd), Events: unix.POLLIN})

			n, err := unix.Poll(pollfds, -1)
			if err != nil {
				fmt.Println("[DEBUG] poll error:", err)
				continue
			}
			if n == 0 {
				continue
			}

			for _, pfd := range pollfds {
				// Handle inotify events
				if pfd.Fd == int32(inFd) && pfd.Revents&unix.POLLIN != 0 {
					buf := make([]byte, 4096)
					n, _ := unix.Read(inFd, buf)
					offset := 0
					for offset < n {
						raw := (*unix.InotifyEvent)(unsafe.Pointer(&buf[offset]))
						nameBytes := buf[offset+unix.SizeofInotifyEvent : offset+unix.SizeofInotifyEvent+int(raw.Len)]
						name := string(nameBytes[:len(nameBytes)-1]) // trim null
						path := filepath.Join("/dev/input", name)

						if raw.Mask&unix.IN_CREATE != 0 {
							if filepath.Base(path) == "mice" || filepath.HasPrefix(filepath.Base(path), "mouse") {
								if dev, err := openMouseDevice(path); err == nil {
									devices[path] = dev
								}
							}
						}
						if raw.Mask&unix.IN_DELETE != 0 {
							if dev, ok := devices[path]; ok {
								dev.Close()
								delete(devices, path)
							}
						}
						offset += unix.SizeofInotifyEvent + int(raw.Len)
					}
				}

				// Handle mouse data
				if pfd.Fd != int32(inFd) && pfd.Revents&unix.POLLIN != 0 {
					buf := make([]byte, reportSize)
					n, err := unix.Read(int(pfd.Fd), buf)
					if err != nil || n < reportSize {
						continue
					}

					buttons := []string{}
					if buf[0]&0x1 != 0 {
						buttons = append(buttons, "L")
					}
					if buf[0]&0x2 != 0 {
						buttons = append(buttons, "R")
					}
					if buf[0]&0x4 != 0 {
						buttons = append(buttons, "M")
					}

					out <- MouseEvent{
						Timestamp: time.Now().UnixMilli(),
						Device:    fmt.Sprintf("fd=%d", pfd.Fd),
						Buttons:   buttons,
						DX:        int8(buf[1]),
						DY:        int8(buf[2]),
					}
				}
			}
		}
	}()

	return out
}

File: pkg/input/mouse.py
#!/usr/bin/env python3
import os, time, select, struct, glob

REPORT_SIZE = 3   # /dev/input/mouseX gives 3-byte packets
POLL_INTERVAL = 0.01
HOTPLUG_SCAN_INTERVAL = 2.0   # seconds between rescans

def decode_mice_packet(data: bytes):
    if len(data) < 3:
        return None
    buttons = data[0]
    dx = struct.unpack("b", data[1:2])[0]  # signed 8-bit
    dy = struct.unpack("b", data[2:3])[0]  # signed 8-bit

    pressed = []
    if buttons & 0x1: pressed.append("L")
    if buttons & 0x2: pressed.append("R")
    if buttons & 0x4: pressed.append("M")

    return f"buttons={pressed if pressed else 'None'} dx={dx} dy={dy}"


class MouseDevice:
    def __init__(self, path):
        self.path = path
        self.fd = None
        self.open()

    def open(self):
        try:
            self.fd = os.open(self.path, os.O_RDONLY | os.O_NONBLOCK)
            print(f"[+] Opened {self.path}")
        except Exception as e:
            self.fd = None

    def close(self):
        if self.fd is not None:
            try:
                os.close(self.fd)
            except:
                pass
            print(f"[-] Closed {self.path}")
            self.fd = None

    def fileno(self):
        return self.fd if self.fd is not None else None

    def read_event(self):
        if self.fd is None:
            return None
        try:
            data = os.read(self.fd, REPORT_SIZE)
            if not data:
                return None
            return decode_mice_packet(data)
        except BlockingIOError:
            return None
        except OSError:
            # device vanished
            self.close()
            return None


def monitor_mice():
    devices = {}
    last_scan = 0

    while True:
        now = time.time()
        if now - last_scan > HOTPLUG_SCAN_INTERVAL:
            last_scan = now
            # scan for /dev/input/mouse* and /dev/input/mice
            candidates = glob.glob("/dev/input/mouse*")
            if os.path.exists("/dev/input/mice"):
                candidates.append("/dev/input/mice")

            found = set(candidates)

            # add new devices
            for path in candidates:
                if path not in devices:
                    dev = MouseDevice(path)
                    if dev.fd is not None:
                        devices[path] = dev

            # remove vanished
            for path in list(devices.keys()):
                if path not in found:
                    devices[path].close()
                    del devices[path]

        if devices:
            try:
                rlist, _, _ = select.select(
                    [d.fileno() for d in devices.values() if d.fileno() is not None],
                    [], [], POLL_INTERVAL
                )
                for fd in rlist:
                    dev = next((d for d in devices.values() if d.fileno() == fd), None)
                    if not dev:
                        continue
                    out = dev.read_event()
                    if out:
                        ts = int(time.time()*1000)
                        print(f"[{ts} ms] {os.path.basename(dev.path)}: {out}")
            except Exception:
                pass
        else:
            time.sleep(POLL_INTERVAL)  # avoid busy loop if nothing connected


if __name__ == "__main__":
    try:
        monitor_mice()
    except KeyboardInterrupt:
        print("\nStopped.")
